File: src/hgraph/_impl/_runtime/_nested_evaluation_engine.py
from abc import abstractmethod
from datetime import datetime
from typing import Mapping, Any, Callable

from hgraph._runtime._constants import MIN_DT, MAX_DT, MIN_TD
from hgraph._impl._runtime._node import NodeImpl
from hgraph._runtime._evaluation_clock import EngineEvaluationClock, EngineEvaluationClockDelegate
from hgraph._runtime._evaluation_engine import EvaluationEngine, EvaluationEngineDelegate
from hgraph._runtime._node import NodeSignature

__all__ = (
    "PythonNestedNodeImpl",
    "NestedEvaluationEngine",
    "NestedEngineEvaluationClock",
)


class NestedEngineEvaluationClock(EngineEvaluationClockDelegate):

    def __init__(self, engine_evaluation_clock: EngineEvaluationClock, nested_node: "PythonNestedNodeImpl"):
        super().__init__(engine_evaluation_clock)
        self._nested_node = nested_node
        self._nested_next_scheduled_evaluation_time = MAX_DT

    @property
    def node(self) -> "PythonNestedNodeImpl":
        return self._nested_node

    @property
    def next_scheduled_evaluation_time(self):
        return self._nested_next_scheduled_evaluation_time

    def reset_next_scheduled_evaluation_time(self):
        self._nested_next_scheduled_evaluation_time = MAX_DT

    def update_next_scheduled_evaluation_time(self, next_time: datetime):
        if (let := self._nested_node.last_evaluation_time) and let >= next_time or self._nested_node.is_stopping:
            return

        proposed_next_time = min(next_time, max(self._nested_next_scheduled_evaluation_time, (let or MIN_DT) + MIN_TD))

        if proposed_next_time != self._nested_next_scheduled_evaluation_time:
            self._nested_next_scheduled_evaluation_time = proposed_next_time
            self._nested_node.graph.schedule_node(self._nested_node.node_ndx, proposed_next_time)


class NestedEvaluationEngine(EvaluationEngineDelegate):
    """

    Requesting a stop of the engine will stop the outer engine.
    Stopping an inner graph is a source of bugs and confusion. Instead, the user should create a mechanism to
    remove the key used to create the graph.
    """

    def __init__(self, engine: EvaluationEngine, evaluation_clock: EngineEvaluationClock):
        super().__init__(engine)
        self._engine_evaluation_clock = evaluation_clock
        self._nested_start_time = evaluation_clock.evaluation_time

    @property
    def start_time(self) -> datetime:
        return self._nested_start_time

    @property
    def evaluation_clock(self) -> "EvaluationClock":
        return self._engine_evaluation_clock

    @property
    def engine_evaluation_clock(self) -> "EngineEvaluationClock":
        return self._engine_evaluation_clock


class PythonNestedNodeImpl(NodeImpl):

    def __init__(
        self,
        node_ndx: int,
        owning_graph_id: tuple[int, ...],
        signature: NodeSignature,
        scalars: Mapping[str, Any],
        eval_fn: Callable = None,
        start_fn: Callable = None,
        stop_fn: Callable = None,
    ):
        super().__init__(node_ndx, owning_graph_id, signature, scalars, eval_fn, start_fn, stop_fn)
        self._last_evaluation_time = None

    @property
    def last_evaluation_time(self) -> datetime:
        return self._last_evaluation_time

    def mark_evaluated(self):
        self._last_evaluation_time = self.graph.evaluation_clock.evaluation_time

    @abstractmethod
    def nested_graphs(self): ...

File: cpp/src/cpp/nodes/nested_evaluation_engine.cpp
#include <hgraph/nodes/nested_evaluation_engine.h>
#include <hgraph/nodes/nested_node.h>
#include <hgraph/types/graph.h>

#include <utility>

namespace hgraph {
    NestedEngineEvaluationClock::NestedEngineEvaluationClock(EngineEvaluationClock::ptr engine_evaluation_clock,
                                                             nested_node_ptr nested_node)
        : EngineEvaluationClockDelegate(std::move(engine_evaluation_clock)), _nested_node(nested_node) {
    }

    nested_node_ptr NestedEngineEvaluationClock::node() const { return _nested_node; }

    engine_time_t NestedEngineEvaluationClock::next_scheduled_evaluation_time() const {
        return _nested_next_scheduled_evaluation_time;
    }

    void NestedEngineEvaluationClock::reset_next_scheduled_evaluation_time() {
        _nested_next_scheduled_evaluation_time = MAX_DT;
    }

    void NestedEngineEvaluationClock::update_next_scheduled_evaluation_time(engine_time_t next_time) {
        auto let{_nested_node->last_evaluation_time()};
        auto eval_time = evaluation_time();

        // Debug output
        // std::cout << "update_next_scheduled_evaluation_time: next_time=" << next_time.time_since_epoch().count()
        //           << " let=" << let.time_since_epoch().count()
        //           << " eval_time=" << eval_time.time_since_epoch().count() << std::endl;

        //Unlike python when not set let will be MIN_DT
        // Python: if (let := self._nested_node.last_evaluation_time) and let >= next_time or self._nested_node.is_stopping:
        // In python we evaluate left to right so putting the brackets on the left side should be consistent
        if ((let != MIN_DT /* equivalent to falisy */ && let >= next_time) || _nested_node->is_stopping()) { return; }

        // Match Python: min(next_time, max(self._nested_next_scheduled_evaluation_time, (let or MIN_DT) + MIN_TD))
        // Note let or MIN_DT is equivalent to let
        // CRITICAL FIX: Also ensure we never schedule before current evaluation time
        auto min_allowed_time = std::max(eval_time, let + MIN_TD);
        auto proposed_next_time = std::min(
            next_time, std::max(_nested_next_scheduled_evaluation_time, min_allowed_time));

        if (proposed_next_time != _nested_next_scheduled_evaluation_time) {
            _nested_next_scheduled_evaluation_time = proposed_next_time;
            _nested_node->graph()->schedule_node(_nested_node->node_ndx(), proposed_next_time);
        }
    }

    NestedEvaluationEngine::NestedEvaluationEngine(EvaluationEngine::s_ptr engine,
                                                   EngineEvaluationClock::s_ptr evaluation_clock)
        : EvaluationEngineDelegate(std::move(engine)), _engine_evaluation_clock(std::move(evaluation_clock)),
          _nested_start_time(_engine_evaluation_clock->evaluation_time()) {
    }

    engine_time_t NestedEvaluationEngine::start_time() const { return _nested_start_time; }

    EvaluationClock::s_ptr NestedEvaluationEngine::evaluation_clock() { return _engine_evaluation_clock; }

    const EngineEvaluationClock::s_ptr& NestedEvaluationEngine::engine_evaluation_clock() { return _engine_evaluation_clock; }

} // namespace hgraph